### PR TITLE
ceph-volume: assign seastore as object store backend when deplying crimson-osd using LVM with cephadm

### PR DESCRIPTION
**Pull request by Svelar**
The only way we can use SeaStore now is vstart shell. And images offered by 'quay.ceph.io' don't include version with crimson+ SeaStore. This PR is about deploying a 'real' cluster with cephadm to utilize and test SeaStore.

After starting a cluster:

write yaml file and apply, eg. `vi osd.yaml` 
```
service_type: osd
service_id: a
placement:
service_type: osd
service_id: a
placement:
  hosts:
    - node2
osds_per_device: 2
spec:
  data_devices:
    paths:
    - /dev/nvme3n1
  objectstore: seastore
```
```
ceph orch apply -i osd.yaml
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by adk3798** [comment]
@Svelar typically for the PR title and commit we start with the component being changed. Even though the change is about seastore/crimson, we're really changing mostly ceph-volume and a bit of cephadm here, so I've adjusted the title appropriately.

File: src/cephadm/cephadm.py
Comment (adk3798) [inline src/cephadm/cephadm.py:3516]: Using the `/etc/ceph/ceph.conf` on the host for deciding the objectstore to use is sort of going away from how cpehadm typically handles these things. I think it would be better if this was added as an option to the `objectstore` parameter for OSD specs (https://docs.ceph.com/en/quincy/cephadm/services/osd/#ceph.deployment.drive_group.DriveGroupSpec.objectstore) and then passed as some kind of argument from when the cephadm mgr module makes the deploy call (in `_create_daemon` in serve.py https://github.com/ceph/ceph/blob/main/src/pybind/mgr/cephadm/serve.py#L1253). It would require having `_create_daemon` function check for it when it's handling an OSD and probably require adding a `--objectstore` flag to the cephadm binary `deploy` command as well.

File: src/python-common/ceph/deployment/translate.py
Comment (adk3798) [inline src/python-common/ceph/deployment/translate.py:188]: this block here implies we won't need to ever add things like `--db-devices` or `--wal-devices` when using seastore (they're only in the bluestore block above). Is that correct?

Comment (Svelar) [reply src/python-common/ceph/deployment/translate.py:188]: Yes.

File: src/cephadm/cephadm.py
Comment (Svelar) [reply src/cephadm/cephadm.py:3516]: Thanks for the comment, the thought is pretty good. But crimson needs to read local_conf, it means that I could change to get objectstore from spec instead of conf file, but I still need to use 'set-extra-ceph-conf' to change conf file, is it OK?

Comment (Svelar) [reply src/cephadm/cephadm.py]: Hi, I have modified, please review.

**Comment by Svelar** [comment]
jenkins retest this please

File: src/cephadm/cephadm.py
Comment (adk3798) [reply src/cephadm/cephadm.py]: As for the conf crimson reads, I expect it will be the one inside the OSD container and not the one actually at /etc/ceph/ceph.conf on the host. For deploying OSDs priort to your patch we already include a conf file at /var/lib/ceph/<fsid>/osd.<osd-id>/config and then mount it at /etc/ceph/ceph.conf in the container, so I think these new OSDs should be able to work the same way.

File: src/pybind/mgr/cephadm/serve.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/serve.py:1268]: I think this should be handled in the other block we already have for adding additional args for osd deployments. A bit above here there is a block
```
# osd deployments needs an --osd-uuid arg
if daemon_spec.daemon_type == 'osd':
    if not osd_uuid_map:
        osd_uuid_map = self.mgr.get_osd_uuid_map()
    osd_uuid = osd_uuid_map.get(daemon_spec.daemon_id)
    if not osd_uuid:
        raise OrchestratorError('osd.%s not in osdmap' % daemon_spec.daemon_id)
    daemon_spec.extra_args.extend(['--osd-fsid', osd_uuid])
```
If there, you just add something like
```
spec = cast(DriveGroupSpec, self.mgr.spec_store[daemon_spec.service_name].spec)
objectstore = spec.objectstore
if objectstore:
    daemon_spec.extra_args.extend(['--objectstore', objectstore])
```
that should put the arg in the correct spot (note that `--osd-fsid` is another arg for the deploy command specifically and we do it the same way by adding it to `daemon_sepc.extra_args`) and it will keep all of our osd specific changes in one location.

File: src/pybind/mgr/cephadm/serve.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/serve.py:1553]: this part seems to be undoing a change from a recent PR. Was this on purpose? If so, what's the reasoning?

**Comment by adk3798** [comment]
want to also note in my reviews I'm only reviewing the cephadm portion of the changes here. The ceph-volume related work will need reviews from others with experience there.

File: src/cephadm/cephadm.py
Comment (adk3798) [reply src/cephadm/cephadm.py]: the ceph conf the osd mounts is generated by cephadm so in theory we should be able to modify it if need be as part of the osd deployment. I think this would require implementing a `generate_config` method for the OSDService class (pybind/mgr/cephadm/osd.py). That function would probably have to call the parent class version (`super().generate_config()`) and then add the objectstore setting to to the ceph conf that is generated (possibly making use of the work from https://github.com/ceph/ceph/pull/50318/commits/0d1596e9275e8ca1b8c09b1490b3e595130aa97b)

Comment (Svelar) [reply src/cephadm/cephadm.py]: Thanks for your help. I choose to do these all in serve.py, because in my sample yaml file, 'objectstore' is under spec section which is DriveGroupSpec, it's easier to handle all operation in a place. For now, no need to use `set-extra-ceph-conf`, only apply yaml.

File: src/pybind/mgr/cephadm/serve.py
Comment (Svelar) [reply src/pybind/mgr/cephadm/serve.py]: Thanks, have modified.

File: src/pybind/mgr/cephadm/serve.py
Comment (Svelar) [reply src/pybind/mgr/cephadm/serve.py]: Apologize for my mistake. Redo this already.

**Comment by Svelar** [comment]
@guits help review 'ceph-volume' part please, thanks.

**Comment by Svelar** [comment]
jenkins retest this please

**Comment by Svelar** [comment]
jenkins retest this please

File: src/pybind/mgr/cephadm/serve.py
Comment (Svelar) [reply src/pybind/mgr/cephadm/serve.py]: But there is a weird thing, it can't pass 'make check'. It shows that 'Incompatible types in assignment (expression has type "DriveGroupSpec", variable has type "CustomContainerSpec")' in the Jenkins report, IMO osd daemon should not have CustomContainerSpec.

Comment (adk3798) [reply src/pybind/mgr/cephadm/serve.py]: I checked and it seems to be taking issue with us reusing the variable named `spec`. I made a sample commit that gets mypy to pass by renaming the variables to differentiate them so you'll probably need something similar https://github.com/adk3798/ceph/commit/4c55b5a38851b23a1c9f131dd4f136828cd170b8

Comment (Svelar) [reply src/pybind/mgr/cephadm/serve.py]: Sorry, I boobed, it is happy now.

**Review by athanatos — APPROVED** [review]
This seems reasonable to me from a crimson/seastore standpoint, though I'm not all that familiar with the ceph-volume or cephadm implementations.

**Comment by Svelar** [comment]
Any more reviews/ comments?

**Comment by guits** [comment]
@Svelar This shouldn't have been merged without explicit approval from a ceph-volume maintainer. As far as I know, you are not a maintainer of ceph-volume.

This might need to be reverted until a proper review is made.

**Pull request merged**
